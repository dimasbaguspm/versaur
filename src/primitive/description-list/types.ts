import type { ReactNode, HTMLAttributes } from 'react'
import type { TextProps } from '../text'

/**
 * Props for the DescriptionList root component
 */
export interface DescriptionListProps extends HTMLAttributes<HTMLDListElement> {
  /**
   * Children should be DescriptionList.Item elements
   */
  children: ReactNode
}

/**
 * Props for DescriptionList.Item
 */
export interface DescriptionListItemProps
  extends HTMLAttributes<HTMLDivElement> {
  /**
   * Children should be DescriptionList.Term and DescriptionList.Details
   */
  children: ReactNode
  /**
   * Number of grid columns to span (default: 4)
   */
  span?: number
}

/**
 * Props for DescriptionList.Term
 */
export interface DescriptionListTermProps
  extends Omit<HTMLAttributes<HTMLElement>, 'children'> {
  /**
   * Term label (required, string only)
   */
  children: string
}

/**
 * Props for DescriptionList.Details
 */
export interface DescriptionListDetailsProps
  extends Omit<HTMLAttributes<HTMLElement>, 'children'> {
  /**
   * Details value (required, string only)
   */
  children: string
  /**
   * Text Color (default: 'black')
   */
  color?: TextProps['color']
}
