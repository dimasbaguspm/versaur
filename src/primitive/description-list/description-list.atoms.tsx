import { forwardRef } from 'react'
import { cn } from '@/utils/cn'
import type {
  DescriptionListItemProps,
  DescriptionListTermProps,
  DescriptionListDetailsProps,
} from './types'
import { getColSpan } from './helpers'
import { Text } from '../text'

/**
 * DescriptionList.Item atom
 */
export const DescriptionListItem = forwardRef<
  HTMLDivElement,
  DescriptionListItemProps
>(function DescriptionListItem(
  { children, className, span = 4, ...props },
  ref
) {
  // Always use col-span-4 by default, allow override via span prop (1-12)
  const colSpan = getColSpan(span)
  return (
    <div ref={ref} className={cn(colSpan, className)} {...props}>
      {children}
    </div>
  )
})

/**
 * DescriptionList.Term atom
 */
export const DescriptionListTerm = forwardRef<
  HTMLElement,
  DescriptionListTermProps
>(function DescriptionListTerm({ children, ...props }, ref) {
  return (
    <dt ref={ref} {...props}>
      <Text as='span' fontSize='sm' fontWeight='medium' color='gray'>
        {children}
      </Text>
    </dt>
  )
})

/**
 * DescriptionList.Details atom
 */
export const DescriptionListDetails = forwardRef<
  HTMLElement,
  DescriptionListDetailsProps
>(function DescriptionListDetails(
  { children, color = 'black', ...props },
  ref
) {
  return (
    <dd ref={ref} {...props}>
      <Text as='span' fontSize='base' fontWeight='medium' color={color}>
        {children}
      </Text>
    </dd>
  )
})
