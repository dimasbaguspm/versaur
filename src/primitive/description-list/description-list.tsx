import { forwardRef } from 'react'
import { cn } from '@/utils/cn'
import type { DescriptionListProps } from './types'
import {
  DescriptionListItem,
  DescriptionListTerm,
  DescriptionListDetails,
} from './description-list.atoms'

/**
 * Compound DescriptionList component for Versaur UI
 *
 * Usage:
 * <DescriptionList>
 *   <DescriptionList.Item>
 *     <DescriptionList.Term>Foo</DescriptionList.Term>
 *     <DescriptionList.Details>Bar</DescriptionList.Details>
 *   </DescriptionList.Item>
 * </DescriptionList>
 */
const DescriptionListRoot = forwardRef<HTMLDListElement, DescriptionListProps>(
  function DescriptionListRoot({ children, className, ...props }, ref) {
    // Always use grid-cols-12 for the root
    return (
      <dl
        ref={ref}
        className={cn('grid grid-cols-12 gap-y-4', className)}
        {...props}
      >
        {children}
      </dl>
    )
  }
)

export const DescriptionList = Object.assign(DescriptionListRoot, {
  Item: DescriptionListItem,
  Term: DescriptionListTerm,
  Details: DescriptionListDetails,
})
