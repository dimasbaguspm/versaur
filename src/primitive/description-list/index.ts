export { DescriptionList } from './description-list'
export * from './types'
