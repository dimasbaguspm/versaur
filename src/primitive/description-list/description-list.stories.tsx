/**
 * DescriptionList stories for Versaur UI
 *
 * Demonstrates semantic usage, span prop, and styled terms/details
 */
import type { Meta, StoryObj } from '@storybook/react'
import { DescriptionList } from './description-list'

const meta: Meta<typeof DescriptionList> = {
  title: 'Primitive/DescriptionList',
  component: DescriptionList,
}
export default meta

type Story = StoryObj<typeof DescriptionList>

export const Basic: Story = {
  render: () => (
    <DescriptionList>
      <DescriptionList.Item>
        <DescriptionList.Term>Name</DescriptionList.Term>
        <DescriptionList.Details>Jane Doe</DescriptionList.Details>
      </DescriptionList.Item>
      <DescriptionList.Item>
        <DescriptionList.Term>Email</DescriptionList.Term>
        <DescriptionList.Details>jane@example.com</DescriptionList.Details>
      </DescriptionList.Item>
      <DescriptionList.Item>
        <DescriptionList.Term>Role</DescriptionList.Term>
        <DescriptionList.Details>Administrator</DescriptionList.Details>
      </DescriptionList.Item>
    </DescriptionList>
  ),
}

export const WithSpan: Story = {
  render: () => (
    <DescriptionList>
      <DescriptionList.Item span={6}>
        <DescriptionList.Term>Project</DescriptionList.Term>
        <DescriptionList.Details>Versaur UI</DescriptionList.Details>
      </DescriptionList.Item>
      <DescriptionList.Item span={3}>
        <DescriptionList.Term>Status</DescriptionList.Term>
        <DescriptionList.Details>Active</DescriptionList.Details>
      </DescriptionList.Item>
      <DescriptionList.Item span={3}>
        <DescriptionList.Term>Contributors</DescriptionList.Term>
        <DescriptionList.Details>5</DescriptionList.Details>
      </DescriptionList.Item>
    </DescriptionList>
  ),
}

export const CustomColors: Story = {
  render: () => (
    <DescriptionList>
      <DescriptionList.Item>
        <DescriptionList.Term>Primary</DescriptionList.Term>
        <DescriptionList.Details color='primary'>Coral</DescriptionList.Details>
      </DescriptionList.Item>
      <DescriptionList.Item>
        <DescriptionList.Term>Secondary</DescriptionList.Term>
        <DescriptionList.Details color='secondary'>
          Sage
        </DescriptionList.Details>
      </DescriptionList.Item>
      <DescriptionList.Item>
        <DescriptionList.Term>Tertiary</DescriptionList.Term>
        <DescriptionList.Details color='tertiary'>Mist</DescriptionList.Details>
      </DescriptionList.Item>
    </DescriptionList>
  ),
}
