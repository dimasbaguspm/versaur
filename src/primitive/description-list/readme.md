# DescriptionList

A semantic, accessible, and flexible description list component for Versaur UI, supporting grid layout, custom column span, and styled terms/details.

## Features
- Renders a semantic `<dl>` with terms (`<dt>`) and descriptions (`<dd>`) using a compound component API
- Grid-based layout with 12 columns for advanced alignment
- Each item can span 1–12 columns (default: 4)
- Terms and details are always strings for clarity and accessibility
- Details support color prop for semantic styling
- Mobile-first, accessible, and themeable with Tailwind and Versaur color system

## Usage
```tsx
import { DescriptionList } from '@/primitive/description-list'

<DescriptionList>
  <DescriptionList.Item>
    <DescriptionList.Term>Name</DescriptionList.Term>
    <DescriptionList.Details>Jane Doe</DescriptionList.Details>
  </DescriptionList.Item>
  <DescriptionList.Item span={6}>
    <DescriptionList.Term>Email</DescriptionList.Term>
    <DescriptionList.Details color="primary">jane@example.com</DescriptionList.Details>
  </DescriptionList.Item>
</DescriptionList>
```

## API

### `<DescriptionList>`
- **Props:**
  - `children`: Only accepts `DescriptionList.Item` elements
  - All standard `<dl>` props
  - Always uses `grid-cols-12` for layout (not user-configurable)

### `<DescriptionList.Item>`
- **Props:**
  - `span` (number, optional): Number of columns to span (1–12, default: 4)
  - `children`: One `DescriptionList.Term` and one `DescriptionList.Details`

### `<DescriptionList.Term>`
- **Props:**
  - `children` (string, required): The term/label

### `<DescriptionList.Details>`
- **Props:**
  - `children` (string, required): The value/description
  - `color` (string, optional): Semantic color (e.g., 'primary', 'secondary', 'tertiary', etc.)

## Example
```tsx
<DescriptionList>
  <DescriptionList.Item span={6}>
    <DescriptionList.Term>Project</DescriptionList.Term>
    <DescriptionList.Details color="primary">Versaur UI</DescriptionList.Details>
  </DescriptionList.Item>
  <DescriptionList.Item span={3}>
    <DescriptionList.Term>Status</DescriptionList.Term>
    <DescriptionList.Details color="success">Active</DescriptionList.Details>
  </DescriptionList.Item>
  <DescriptionList.Item span={3}>
    <DescriptionList.Term>Contributors</DescriptionList.Term>
    <DescriptionList.Details>5</DescriptionList.Details>
  </DescriptionList.Item>
</DescriptionList>
```

## Accessibility
- Uses semantic HTML (`<dl>`, `<dt>`, `<dd>`) for screen reader support
- Requires string children for terms and details for clarity
- Fully keyboard accessible

## See Also
- See `types.ts` for full prop documentation.
- See Storybook for live examples and visual documentation.
