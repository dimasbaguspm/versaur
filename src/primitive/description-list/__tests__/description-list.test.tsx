import { render, screen } from '@testing-library/react'
import { DescriptionList } from '../description-list'

describe('DescriptionList', () => {
  it('renders correctly (snapshot)', () => {
    const { asFragment } = render(
      <DescriptionList>
        <DescriptionList.Item>
          <DescriptionList.Term>Name</DescriptionList.Term>
          <DescriptionList.Details>Jane Doe</DescriptionList.Details>
        </DescriptionList.Item>
        <DescriptionList.Item span={6}>
          <DescriptionList.Term>Email</DescriptionList.Term>
          <DescriptionList.Details color='primary'>
            jane@example.com
          </DescriptionList.Details>
        </DescriptionList.Item>
      </DescriptionList>
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('renders all terms and details as strings', () => {
    render(
      <DescriptionList>
        <DescriptionList.Item>
          <DescriptionList.Term>Name</DescriptionList.Term>
          <DescriptionList.Details>Jane Doe</DescriptionList.Details>
        </DescriptionList.Item>
        <DescriptionList.Item>
          <DescriptionList.Term>Email</DescriptionList.Term>
          <DescriptionList.Details>jane@example.com</DescriptionList.Details>
        </DescriptionList.Item>
      </DescriptionList>
    )
    expect(screen.getByText('Name')).toBeInTheDocument()
    expect(screen.getByText('Jane Doe')).toBeInTheDocument()
    expect(screen.getByText('Email')).toBeInTheDocument()
    expect(screen.getByText('jane@example.com')).toBeInTheDocument()
  })

  it('applies the correct col-span class for span prop', () => {
    render(
      <DescriptionList>
        <DescriptionList.Item span={6}>
          <DescriptionList.Term>Project</DescriptionList.Term>
          <DescriptionList.Details>Versaur UI</DescriptionList.Details>
        </DescriptionList.Item>
      </DescriptionList>
    )
    const item = screen.getByText('Project').closest('div')
    expect(item).toHaveClass('col-span-6')
  })

  it('applies the correct color to details', () => {
    render(
      <DescriptionList>
        <DescriptionList.Item>
          <DescriptionList.Term>Status</DescriptionList.Term>
          <DescriptionList.Details color='success'>
            Active
          </DescriptionList.Details>
        </DescriptionList.Item>
      </DescriptionList>
    )
    expect(screen.getByText('Active')).toBeInTheDocument()
  })
})
