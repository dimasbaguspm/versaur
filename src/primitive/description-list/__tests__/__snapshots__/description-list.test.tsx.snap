// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DescriptionList > renders correctly (snapshot) 1`] = `
<DocumentFragment>
  <dl
    class="grid grid-cols-12 gap-y-4"
  >
    <div
      class="col-span-4"
    >
      <dt>
        <span
          class="text-gray-500 text-left text-sm font-medium"
        >
          Name
        </span>
      </dt>
      <dd>
        <span
          class="text-black text-left text-base font-medium"
        >
          Jane Doe
        </span>
      </dd>
    </div>
    <div
      class="col-span-6"
    >
      <dt>
        <span
          class="text-gray-500 text-left text-sm font-medium"
        >
          Email
        </span>
      </dt>
      <dd>
        <span
          class="text-primary text-left text-base font-medium"
        >
          jane@example.com
        </span>
      </dd>
    </div>
  </dl>
</DocumentFragment>
`;
